# MM&A Supermarket Case

## 1. Exploring the Data (EDA)

In [1]:
# importing necessary libraries
import numpy as np
import pandas as pd

In [2]:
# reading in the csv file into a dataframe
file = 'mma_mart.csv'
df = pd.read_csv(file)

In [3]:
# observing the data structure
df.head()

order_id  product_id                                   product_name  \
0         1       49302                               Bulgarian Yogurt   
1         1       11109  Organic 4% Milk Fat Whole Milk Cottage Cheese   
2         1       10246                          Organic Celery Hearts   
3         1       49683                                 Cucumber Kirby   
4         1       43633           Lightly Smoked Sardines in Olive Oil   

   aisle_id                 aisle  department_id    department  
0       120                yogurt             16    dairy eggs  
1       108  other creams cheeses             16    dairy eggs  
2        83      fresh vegetables              4       produce  
3        83      fresh vegetables              4       produce  
4        95   canned meat seafood             15  canned goods

In [4]:
# finding the shape of the data set (row and column)
df.shape

(987259, 7)

The data shows 987259 rows and 7 columns.

In [5]:
# finding amount of unique entries in each column
df.nunique()

order_id         97833
product_id       35070
product_name     35070
aisle_id           134
aisle              134
department_id       21
department          21
dtype: int64

There are ~98000 orders and ~35000 unique products which is across 134 aisles that belong to 21 departments. This shows that selecting 1000-1200 products for the instabasket aisle is selecting only 1000/35000 = ~3% of the total product selection offered at MM&A supermarket.

Next, exploring the aisle and departments to maybe gain some insight on what products should be considered as refrigerant and frozen products.

In [6]:
df['aisle'].unique()

array(['yogurt', 'other creams cheeses', 'fresh vegetables',
       'canned meat seafood', 'fresh fruits', 'packaged cheese', 'eggs',
       'spices seasonings', 'oils vinegars', 'baking ingredients',
       'doughs gelatins bake mixes', 'spreads',
       'packaged vegetables fruits', 'soy lactosefree', 'poultry counter',
       'bread', 'breakfast bakery', 'cold flu allergy',
       'energy granola bars', 'breakfast bars pastries', 'chips pretzels',
       'trail mix snack mix', 'crackers', 'refrigerated',
       'energy sports drinks', 'salad dressing toppings',
       'prepared soups salads', 'milk', 'paper goods',
       'water seltzer sparkling water', 'kosher foods',
       'packaged poultry', 'instant foods', 'packaged produce',
       'cookies cakes', 'candy chocolate', 'body lotions soap',
       'dry pasta', 'laundry', 'air fresheners candles', 'frozen produce',
       'buns rolls', 'canned fruit applesauce', 'juice nectars',
       'granola', 'fresh herbs', 'baby food formul

As can be seen from the aisle names, there are aisles with "frozen" such as `frozen produce` or `frozen meat seafood` in it which all products listed under these aisles would be considered for the 100 frozen product for instabasket. The same can't be said for refridgerant as there a lot of aisles to parse through so it might be better to look at the department to determine what department contains refridgerant products.

In [7]:
df['department'].unique()

array(['dairy eggs', 'produce', 'canned goods', 'pantry', 'meat seafood',
       'bakery', 'personal care', 'snacks', 'breakfast', 'beverages',
       'deli', 'household', 'international', 'dry goods pasta', 'frozen',
       'babies', 'pets', 'alcohol', 'bulk', 'missing', 'other'],
      dtype=object)

From the look at department names, `dairy eggs`, `produce`, `meat seafood`, `deli`, `frozen`, `missing` and `other` are all potential products for fridge and freezer space so it might be better to take a closer look at the products in those department.

In [8]:
dept_list = ['dairy eggs', 'produce', 'meat seafood', 'deli', 'frozen', 'missing', 'other']

In [9]:
# looking at products and aisles of those products for insight as well as how many products each department had
for department in dept_list:
    print(f"Department: {department}")
    print(f"Number of products: {df[df['department']==department]['product_name'].nunique()}\n")
    display(df[df['department']==department][['product_name', 'aisle']])

Department: dairy eggs
Number of products: 2886



product_name  \
0                                        Bulgarian Yogurt   
1           Organic 4% Milk Fat Whole Milk Cottage Cheese   
7                             Organic Whole String Cheese   
8                                      Organic Egg Whites   
17      Total 2% with Strawberry Lowfat Greek Strained...   
...                                                   ...   
987240                        Large Grade AA Organic Eggs   
987241                Reduced Fat Mozarella String Cheese   
987242                   Vanilla Light & Fit Greek Yogurt   
987243               Non-Fat Vanilla Blended Greek Yogurt   
987254            Total 2% All Natural Plain Greek Yogurt   

                       aisle  
0                     yogurt  
1       other creams cheeses  
7            packaged cheese  
8                       eggs  
17                    yogurt  
...                      ...  
987240                  eggs  
987241       packaged cheese  
987242                yogurt  
987243                yogurt  
987254                yogurt  

[164468 rows x 2 columns]

Department: produce
Number of products: 1437



product_name                       aisle
2           Organic Celery Hearts            fresh vegetables
3                  Cucumber Kirby            fresh vegetables
5          Bag of Organic Bananas                fresh fruits
6            Organic Hass Avocado                fresh fruits
9           Michigan Organic Kale            fresh vegetables
...                           ...                         ...
987245                Gala Apples                fresh fruits
987246       Organic Yellow Onion            fresh vegetables
987247       Organic Baby Carrots  packaged vegetables fruits
987249       Organic Baby Spinach  packaged vegetables fruits
987250  Garnet Sweet Potato (Yam)            fresh vegetables

[288775 rows x 2 columns]

Department: meat seafood
Number of products: 692



product_name  \
23      Air Chilled Organic Boneless Skinless Chicken ...   
46               Boneless Skinless Chicken Breast Fillets   
97                            Boneless Beef Sirloin Steak   
176                                      Low Sodium Bacon   
204                  Boneless And Skinless Chicken Breast   
...                                                   ...   
986977                  Organic Air Chilled Whole Chicken   
987067  Natural Hickory Smoked Canadian Bacon Center C...   
987068  All Natural Boneless & Skinless Chicken Breast...   
987175                  Organic Air Chilled Whole Chicken   
987179                      Organic Beef Uncured Hot Dogs   

                         aisle  
23             poultry counter  
46            packaged poultry  
97                meat counter  
176     hot dogs bacon sausage  
204            poultry counter  
...                        ...  
986977         poultry counter  
987067  hot dogs bacon sausage  
987068        packaged poultry  
987175         poultry counter  
987179  hot dogs bacon sausage  

[21588 rows x 2 columns]

Department: deli
Number of products: 1069



product_name                   aisle
40                           Fresh Fruit Salad   prepared soups salads
101                      Mango Pineapple Salsa    fresh dips tapenades
179     Basil, Asiago & Pine Nut Pesto Ravioli          prepared meals
191                             Yuba Tofu Skin  tofu meat alternatives
192                          Organic Firm Tofu  tofu meat alternatives
...                                        ...                     ...
987153                    Spicy Avocado Hummus    fresh dips tapenades
987158                               Guacamole    fresh dips tapenades
987199                           Guacamole Dip    fresh dips tapenades
987204                         Original Hummus    fresh dips tapenades
987253                         Original Hummus    fresh dips tapenades

[32008 rows x 2 columns]

Department: frozen
Number of products: 3127



product_name                    aisle
68                                   Pineapple Chunks           frozen produce
100            Teriyaki & Pineapple Chicken Meatballs             frozen meals
114     All Natural Boneless Skinless Chicken Breasts      frozen meat seafood
118                           Combination Pizza Rolls  frozen appetizers sides
135                    Organic Mini Homestyle Waffles         frozen breakfast
...                                               ...                      ...
987176                 Organic Ice Cream Vanilla Bean            ice cream ice
987180   Dairy Free Coconut Milk Frozen Dessert Minis            ice cream ice
987184                           Organic Mango Chunks           frozen produce
987210                            Bag of Large Lemons      frozen meat seafood
987215                              Vanilla Ice Cream            ice cream ice

[68202 rows x 2 columns]

Department: missing
Number of products: 518



product_name    aisle
654                          Tomato Basil Bisque Soup  missing
1511      Cold Pressed Watermelon & Lemon Juice Blend  missing
1512                           Paleo Blueberry Muffin  missing
2126    Magic Tape Refillable Dispenser 3/4\" x 850\"  missing
4121                           Organic Poblano Pepper  missing
...                                               ...      ...
986067                       Oatneal Cookie Ice Cream  missing
986086    Dairy Free Unsweetened Almond Milk Beverage  missing
986103                Organic Asian Chopped Salad Kit  missing
986275                                      Lemon Bag  missing
986276      Organic Nondairy Strawberry Cashew Yogurt  missing

[2289 rows x 2 columns]

Department: other
Number of products: 303



product_name  aisle
691            Coffee Mate French Vanilla Creamer Packets  other
1077                        SleepGels Nighttime Sleep Aid  other
1926                             Roasted Unsalted Almonds  other
1985                         Camilia, Single Liquid Doses  other
2127    Maximum Strength Original Paste Diaper Rash Oi...  other
...                                                   ...    ...
983992                           Roasted Unsalted Almonds  other
984112                           Boneless Pork Tenderloin  other
984515                              Roasted Almond Butter  other
984534            Light CocoWhip! Coconut Whipped Topping  other
986326                                   93/7 Ground Beef  other

[1097 rows x 2 columns]

Looking at the results, products from `dairy eggs`, `meat seafood` `deli`, and some products from `missing` and `other` (would need to look at what exactly the product is since there was "Oatneal Cookie Ice Cream" in `missing` aisle which is a frozen item and "Boneless Pork Tenderloin" in `other` which is a `meat seafood` item) would be the departments that require refridgerant. Produce might not be a neccisity to be refridgerated as it can be stored at room temperature but only refridgerated to prolong shelf life so if there is a lot of demand for the produce, then it wouldn't be on the shelf for long but if there is low demand for it, then it might be better to not even have it as a selection so it might be a better idea to give the refridgerator space to the other aisles.

In [10]:
# Sanity check to see if aisles such as frozen produce are part of the frozen department or produce department
frozen_aisles = df[df['aisle'].str.contains('frozen', case=False)]['aisle'].unique()
print(frozen_aisles)

['frozen produce' 'frozen meals' 'frozen meat seafood'
 'frozen appetizers sides' 'frozen breakfast' 'frozen breads doughs'
 'frozen vegan vegetarian' 'frozen pizza' 'frozen dessert' 'frozen juice']


In [11]:
frozen_department = df[df['aisle'].str.contains('frozen', case=False)]['department'].unique()
print(frozen_department)

['frozen']


It is safe to assume that 100 products for the frozen products will be from the frozen department and maybe `missing` and `other` department.

In [12]:
# storing all frozen products into a variable
freezer_df = df[df['department']=='frozen']
freezer_df

order_id  product_id                                   product_name  \
68             7       46802                               Pineapple Chunks   
100           11       30162         Teriyaki & Pineapple Chicken Meatballs   
114           12       38050  All Natural Boneless Skinless Chicken Breasts   
118           12       29471                        Combination Pizza Rolls   
135           14         162                 Organic Mini Homestyle Waffles   
...          ...         ...                                            ...   
987176     99993       43749                 Organic Ice Cream Vanilla Bean   
987180     99993        5537   Dairy Free Coconut Milk Frozen Dessert Minis   
987184     99993       26128                           Organic Mango Chunks   
987210     99996        9434                            Bag of Large Lemons   
987215     99996       32691                              Vanilla Ice Cream   

        aisle_id                    aisle  department_id department  
68           116           frozen produce              1     frozen  
100           38             frozen meals              1     frozen  
114           34      frozen meat seafood              1     frozen  
118          129  frozen appetizers sides              1     frozen  
135           52         frozen breakfast              1     frozen  
...          ...                      ...            ...        ...  
987176        37            ice cream ice              1     frozen  
987180        37            ice cream ice              1     frozen  
987184       116           frozen produce              1     frozen  
987210        34      frozen meat seafood              1     frozen  
987215        37            ice cream ice              1     frozen  

[68202 rows x 7 columns]

In [13]:
# storing all fridge products into a variable
fridge_df = df[df['department'].isin(['dairy eggs', 'meat seafood', 'deli'])]
fridge_df

order_id  product_id  \
0              1       49302   
1              1       11109   
7              1       22035   
8              2       33120   
17             3       33754   
...          ...         ...   
987241     99999       10034   
987242     99999       44085   
987243     99999        9558   
987253    100000       30489   
987254    100000       30169   

                                             product_name  aisle_id  \
0                                        Bulgarian Yogurt       120   
1           Organic 4% Milk Fat Whole Milk Cottage Cheese       108   
7                             Organic Whole String Cheese        21   
8                                      Organic Egg Whites        86   
17      Total 2% with Strawberry Lowfat Greek Strained...       120   
...                                                   ...       ...   
987241                Reduced Fat Mozarella String Cheese        21   
987242                   Vanilla Light & Fit Greek Yogurt       120   
987243               Non-Fat Vanilla Blended Greek Yogurt       120   
987253                                    Original Hummus        67   
987254            Total 2% All Natural Plain Greek Yogurt       120   

                       aisle  department_id  department  
0                     yogurt             16  dairy eggs  
1       other creams cheeses             16  dairy eggs  
7            packaged cheese             16  dairy eggs  
8                       eggs             16  dairy eggs  
17                    yogurt             16  dairy eggs  
...                      ...            ...         ...  
987241       packaged cheese             16  dairy eggs  
987242                yogurt             16  dairy eggs  
987243                yogurt             16  dairy eggs  
987253  fresh dips tapenades             20        deli  
987254                yogurt             16  dairy eggs  

[218064 rows x 7 columns]

In [14]:
# removing the freezer and fridge products from the main df as this will be for the 1000 in aisle product selection
remove_index = freezer_df.index.union(fridge_df.index)
filtered_df = df.drop(remove_index)
filtered_df

order_id  product_id                            product_name  \
2              1       10246                   Organic Celery Hearts   
3              1       49683                          Cucumber Kirby   
4              1       43633    Lightly Smoked Sardines in Olive Oil   
5              1       13176                  Bag of Organic Bananas   
6              1       47209                    Organic Hass Avocado   
...          ...         ...                                     ...   
987252    100000       19508                          Corn Tortillas   
987255    100000       38734                    Wheat Sandwich Thins   
987256    100000       36759  Unscented Long Lasting Stick Deodorant   
987257    100000       37107                            Ground Cumin   
987258    100000       31506                  Extra Virgin Olive Oil   

        aisle_id                 aisle  department_id     department  
2             83      fresh vegetables              4        produce  
3             83      fresh vegetables              4        produce  
4             95   canned meat seafood             15   canned goods  
5             24          fresh fruits              4        produce  
6             24          fresh fruits              4        produce  
...          ...                   ...            ...            ...  
987252       128  tortillas flat bread              3         bakery  
987255       128  tortillas flat bread              3         bakery  
987256        80            deodorants             11  personal care  
987257       104     spices seasonings             13         pantry  
987258        19         oils vinegars             13         pantry  

[700993 rows x 7 columns]

In [15]:
# Sanity Check to see if the rows were removed correctly
freezer_df.shape[0] + fridge_df.shape[0] + filtered_df.shape[0] == df.shape[0]

True

In [16]:
# Check to see if the right departments are removed
filtered_df['department'].unique()

array(['produce', 'canned goods', 'pantry', 'bakery', 'personal care',
       'snacks', 'breakfast', 'beverages', 'household', 'international',
       'dry goods pasta', 'babies', 'pets', 'alcohol', 'bulk', 'missing',
       'other'], dtype=object)

It did remove the aisle correctly. We'll keep `missing` and `other` and see once products are within the top selections of product, if it should be placed in a fridge/freezer. 

## 2. Product Selection

In [17]:
# import the library needed to split
from sklearn.model_selection import train_test_split

# splitting the data 
train_aisle, test_aisle = train_test_split(filtered_df, test_size=0.2, random_state=1)

In [18]:
# Sanity check
train_aisle

order_id  product_id                                   product_name  \
554516     56164       12817      Original Baked Whole Grain Wheat Crackers   
858871     87157       20738                                   Plain Bagels   
536470     54317       36606                     Sprouted Multi-Grain Bread   
770058     78080       36085  Berry Berry Good Lemonade Organic Juice Drink   
86781       8859       21338                                 Silver Tequila   
...          ...         ...                                            ...   
523861     53070        9327                                  Garlic Powder   
692459     70207        1710                             Kombucha Gingerade   
663924     67294       43328                         Ginger Twist Green Tea   
693177     70268       39877                     Organic Granny Smith Apple   
180504     18218        1700                          White Cheddar Popcorn   

        aisle_id              aisle  department_id department  
554516        78           crackers             19     snacks  
858871        93   breakfast bakery              3     bakery  
536470       112              bread              3     bakery  
770058        98      juice nectars              7  beverages  
86781        124            spirits              5    alcohol  
...          ...                ...            ...        ...  
523861       104  spices seasonings             13     pantry  
692459        31       refrigerated              7  beverages  
663924        94                tea              7  beverages  
693177        24       fresh fruits              4    produce  
180504        23      popcorn jerky             19     snacks  

[560794 rows x 7 columns]

In [19]:
# Sanity Check
test_aisle

order_id  product_id                           product_name  aisle_id  \
699855     70941       16759                   Organic Carrot Bunch        83   
128747     13025       16813                  Gluten Free Soft Taco       128   
215472     21758       39945               Magic Eraser Extra Power       114   
795857     80705       37646                    Organic Gala Apples        24   
279647     28200       29479     Gobble Gobble Turkey Pouch Stage 3        92   
...          ...         ...                                    ...       ...   
232063     23448       39408                            Gala Apples        24   
409839     41474       47209                   Organic Hass Avocado        24   
93961       9581       30391                       Organic Cucumber        83   
175346     17669       39429  Organic Large Flour Tortillas 6 Count       128   
845677     85792       14852             All Natural Marinara Sauce         9   

                       aisle  department_id       department  
699855      fresh vegetables              4          produce  
128747  tortillas flat bread              3           bakery  
215472     cleaning products             17        household  
795857          fresh fruits              4          produce  
279647     baby food formula             18           babies  
...                      ...            ...              ...  
232063          fresh fruits              4          produce  
409839          fresh fruits              4          produce  
93961       fresh vegetables              4          produce  
175346  tortillas flat bread              3           bakery  
845677           pasta sauce              9  dry goods pasta  

[140199 rows x 7 columns]

In [20]:
# Checking to see if a product appears multiple time in one order if multiple of the product is ordered 
df[df[['order_id', 'product_id']].duplicated()]

Empty DataFrame
Columns: [order_id, product_id, product_name, aisle_id, aisle, department_id, department]
Index: []

### Method 2: Product Selection Proportional to Aisle Selection

This second method is to get a variety of items from each aisle proportional to the amount of orders for that aisle.

In [21]:
# finding the total amount of products ordered
train_size = train_aisle.shape[0]
train_size

560794

In [23]:
# finding the amount of items ordered for each aisle
aisle_info=train_aisle.groupby('aisle').size().reset_index(name='count').sort_values(by='count', ascending=False)
aisle_info

aisle  count
46                   fresh fruits  88764
49               fresh vegetables  83037
74     packaged vegetables fruits  43464
99  water seltzer sparkling water  20377
24                 chips pretzels  17513
..                            ...    ...
41                   eye ear care    226
60               kitchen supplies    221
2                baby accessories    186
3             baby bath body care    177
8                          beauty    129

[101 rows x 2 columns]

In [24]:
aisle_info['percentage']=aisle_info['count']/train_size*100
aisle_info

aisle  count  percentage
46                   fresh fruits  88764   15.828272
49               fresh vegetables  83037   14.807041
74     packaged vegetables fruits  43464    7.750440
99  water seltzer sparkling water  20377    3.633598
24                 chips pretzels  17513    3.122894
..                            ...    ...         ...
41                   eye ear care    226    0.040300
60               kitchen supplies    221    0.039408
2                baby accessories    186    0.033167
3             baby bath body care    177    0.031562
8                          beauty    129    0.023003

[101 rows x 3 columns]

In [25]:
aisle_info

aisle  count  percentage
46                   fresh fruits  88764   15.828272
49               fresh vegetables  83037   14.807041
74     packaged vegetables fruits  43464    7.750440
99  water seltzer sparkling water  20377    3.633598
24                 chips pretzels  17513    3.122894
..                            ...    ...         ...
41                   eye ear care    226    0.040300
60               kitchen supplies    221    0.039408
2                baby accessories    186    0.033167
3             baby bath body care    177    0.031562
8                          beauty    129    0.023003

[101 rows x 3 columns]

In [31]:
# Creating a mapping to add department into the aisle and count dataframe
train_aisle_mapping = train_aisle[['aisle', 'department']]
train_aisle_mapping = train_aisle_mapping.drop_duplicates()
train_aisle_mapping

aisle     department
554516                    crackers         snacks
858871            breakfast bakery         bakery
536470                       bread         bakery
770058               juice nectars      beverages
86781                      spirits        alcohol
...                            ...            ...
530862                indian foods  international
529399                      beauty  personal care
921313  muscles joints pain relief  personal care
91826             baby accessories         babies
240944                kosher foods  international

[101 rows x 2 columns]

In [32]:
# Merge the 'aisle_count' dataframe with the 'department' column from 'train_aisle' based on the 'aisle' column
merged_df = aisle_info.merge(train_aisle_mapping, on='aisle', how='left')
merged_df

aisle  count  percentage     department
0                     fresh fruits  88764   15.828272        produce
1                 fresh vegetables  83037   14.807041        produce
2       packaged vegetables fruits  43464    7.750440        produce
3    water seltzer sparkling water  20377    3.633598      beverages
4                   chips pretzels  17513    3.122894         snacks
..                             ...    ...         ...            ...
96                    eye ear care    226    0.040300  personal care
97                kitchen supplies    221    0.039408      household
98                baby accessories    186    0.033167         babies
99             baby bath body care    177    0.031562         babies
100                         beauty    129    0.023003  personal care

[101 rows x 4 columns]

Now that the percentage of orders from each aisle has been calculated, the next step would be to see how many items this would translate to for the 800 product aisle.

In [33]:
merged_df['amount'] = round((merged_df['percentage']/100)*800)

In [34]:
merged_df

aisle  count  percentage     department  amount
0                     fresh fruits  88764   15.828272        produce   127.0
1                 fresh vegetables  83037   14.807041        produce   118.0
2       packaged vegetables fruits  43464    7.750440        produce    62.0
3    water seltzer sparkling water  20377    3.633598      beverages    29.0
4                   chips pretzels  17513    3.122894         snacks    25.0
..                             ...    ...         ...            ...     ...
96                    eye ear care    226    0.040300  personal care     0.0
97                kitchen supplies    221    0.039408      household     0.0
98                baby accessories    186    0.033167         babies     0.0
99             baby bath body care    177    0.031562         babies     0.0
100                         beauty    129    0.023003  personal care     0.0

[101 rows x 5 columns]

In [35]:
# Sanity Check to see if it equals 800
merged_df['amount'].sum()

801.0

It did not equal 800 due to rounding so 1 more product needs to be dropped. We'll just be dropping the aisle that is equal to 1 and lowest count.

In [36]:
# finding the aisle with an amount of 1 and least amount of orders
last_row = merged_df[merged_df['amount']==1].iloc[-1]
last_row[0]

'deodorants'

In [37]:
# Changing the amount of the last row to 0
merged_df.loc[merged_df['aisle'] == last_row[0], 'amount'] = 0 

In [38]:
# Sanity Check
new_last_row = merged_df[merged_df['amount']==1].iloc[-1]
new_last_row[0]

'more household'

Now that the amount of products needed for each aisle has been determined, the next step would be to find the top X (X as in amount) product for each aisle as in method 1. 

In [41]:
products_df=train_aisle.groupby('product_id').size().reset_index(name='count').sort_values(by='count', ascending=False)
products_df

product_id  count
12876       24852  11547
6735        13176   9371
10910       21137   6458
11307       21903   5877
24364       47209   5162
...           ...    ...
11778       22773      1
11780       22775      1
1084         2114      1
11781       22778      1
10235       19837      1

[25607 rows x 2 columns]

In [66]:
product_mapping = train_aisle[['product_id', 'product_name', 'aisle_id', 'aisle', 'department_id', 'department']]

In [67]:
product_mapping

product_id                                   product_name  aisle_id  \
554516       12817      Original Baked Whole Grain Wheat Crackers        78   
858871       20738                                   Plain Bagels        93   
536470       36606                     Sprouted Multi-Grain Bread       112   
770058       36085  Berry Berry Good Lemonade Organic Juice Drink        98   
86781        21338                                 Silver Tequila       124   
...            ...                                            ...       ...   
523861        9327                                  Garlic Powder       104   
692459        1710                             Kombucha Gingerade        31   
663924       43328                         Ginger Twist Green Tea        94   
693177       39877                     Organic Granny Smith Apple        24   
180504        1700                          White Cheddar Popcorn        23   

                    aisle  department_id department  
554516           crackers             19     snacks  
858871   breakfast bakery              3     bakery  
536470              bread              3     bakery  
770058      juice nectars              7  beverages  
86781             spirits              5    alcohol  
...                   ...            ...        ...  
523861  spices seasonings             13     pantry  
692459       refrigerated              7  beverages  
663924                tea              7  beverages  
693177       fresh fruits              4    produce  
180504      popcorn jerky             19     snacks  

[560794 rows x 6 columns]

In [81]:
# merged_df2 = products_df.join(train_aisle, how='inner')
# Use the .join() method to join df2 with df1 based on their indices
merged_df2 = products_df.merge(train_aisle, how='right')
merged_df2

product_id  count  order_id  \
0            12817    110     56164   
1            20738    271     87157   
2            36606    141     54317   
3            36085    112     78080   
4            21338     13      8859   
...            ...    ...       ...   
560789        9327    156     53070   
560790        1710     40     70207   
560791       43328      2     67294   
560792       39877   1414     70268   
560793        1700    194     18218   

                                         product_name  aisle_id  \
0           Original Baked Whole Grain Wheat Crackers        78   
1                                        Plain Bagels        93   
2                          Sprouted Multi-Grain Bread       112   
3       Berry Berry Good Lemonade Organic Juice Drink        98   
4                                      Silver Tequila       124   
...                                               ...       ...   
560789                                  Garlic Powder       104   
560790                             Kombucha Gingerade        31   
560791                         Ginger Twist Green Tea        94   
560792                     Organic Granny Smith Apple        24   
560793                          White Cheddar Popcorn        23   

                    aisle  department_id department  
0                crackers             19     snacks  
1        breakfast bakery              3     bakery  
2                   bread              3     bakery  
3           juice nectars              7  beverages  
4                 spirits              5    alcohol  
...                   ...            ...        ...  
560789  spices seasonings             13     pantry  
560790       refrigerated              7  beverages  
560791                tea              7  beverages  
560792       fresh fruits              4    produce  
560793      popcorn jerky             19     snacks  

[560794 rows x 8 columns]

In [82]:
merged_df2

product_id  count  order_id  \
0            12817    110     56164   
1            20738    271     87157   
2            36606    141     54317   
3            36085    112     78080   
4            21338     13      8859   
...            ...    ...       ...   
560789        9327    156     53070   
560790        1710     40     70207   
560791       43328      2     67294   
560792       39877   1414     70268   
560793        1700    194     18218   

                                         product_name  aisle_id  \
0           Original Baked Whole Grain Wheat Crackers        78   
1                                        Plain Bagels        93   
2                          Sprouted Multi-Grain Bread       112   
3       Berry Berry Good Lemonade Organic Juice Drink        98   
4                                      Silver Tequila       124   
...                                               ...       ...   
560789                                  Garlic Powder       104   
560790                             Kombucha Gingerade        31   
560791                         Ginger Twist Green Tea        94   
560792                     Organic Granny Smith Apple        24   
560793                          White Cheddar Popcorn        23   

                    aisle  department_id department  
0                crackers             19     snacks  
1        breakfast bakery              3     bakery  
2                   bread              3     bakery  
3           juice nectars              7  beverages  
4                 spirits              5    alcohol  
...                   ...            ...        ...  
560789  spices seasonings             13     pantry  
560790       refrigerated              7  beverages  
560791                tea              7  beverages  
560792       fresh fruits              4    produce  
560793      popcorn jerky             19     snacks  

[560794 rows x 8 columns]

In [ ]:
for 